// react-native-mock/build/components/ART/Transform.js
var _createClass=function(){function defineProperties(target,props){for(var i=0;i<props.length;i++){var descriptor=props[i];descriptor.enumerable=descriptor.enumerable||false;descriptor.configurable=true;if("value" in descriptor)descriptor.writable=true;Object.defineProperty(target,descriptor.key,descriptor);}}return function(Constructor,protoProps,staticProps){if(protoProps)defineProperties(Constructor.prototype,protoProps);if(staticProps)defineProperties(Constructor,staticProps);return Constructor;};}();function _classCallCheck(instance,Constructor){if(!(instance instanceof Constructor)){throw new TypeError("Cannot call a class as a function");}}var 
Transform=function(){
function Transform(){_classCallCheck(this,Transform);
this.xx=0;
this.yx=0;
this.xy=0;
this.yy=0;
this.x=0;
this.y=0;}_createClass(Transform,[{key:"transformTo",value:function(){function transformTo()

{
return this;}return transformTo;}()},{key:"move",value:function(){function move()

{
return this;}return move;}()},{key:"rotate",value:function(){function rotate()

{
return this;}return rotate;}()},{key:"scale",value:function(){function scale()

{
return this;}return scale;}()},{key:"transform",value:function(){function transform()

{
return this;}return transform;}()}]);return Transform;}();



module.exports=Transform;
